tests/components/mqtt/test_cover.py

```diff
@@ -13,1071 +12,1081 @@
+            'payload_stop': 'STOP'
+        }
+    })
+
+    state = hass.states.get('cover.test')
+    assert STATE_UNKNOWN == state.state
+    assert not state.attributes.get(ATTR_ASSUMED_STATE)
+
+    async_fire_mqtt_message(hass, 'get-position-topic', '0')
+
+    state = hass.states.get('cover.test')
+    assert STATE_CLOSED == state.state
+
+    async_fire_mqtt_message(hass, 'get-position-topic', '100')
+
+    state = hass.states.get('cover.test')
+    assert STATE_OPEN == state.state
+
+
+async def test_state_via_template(hass, mqtt_mock):
+    """Test the controlling state via topic."""
+    assert await async_setup_component(hass, cover.DOMAIN, {
+        cover.DOMAIN: {
+            'platform': 'mqtt',
+            'name': 'test',
+            'state_topic': 'state-topic',
+            'command_topic': 'command-topic',
+            'qos': 0,
+            'value_template': '\
+                {% if (value | multiply(0.01) | int) == 0  %}\
+                  closed\
+                {% else %}\
+                  open\
+                {% endif %}'
+        }
+    })
+
+    state = hass.states.get('cover.test')
+    assert STATE_UNKNOWN == state.state
+
+    async_fire_mqtt_message(hass, 'state-topic', '10000')
+
+    state = hass.states.get('cover.test')
+    assert STATE_OPEN == state.state
+
+    async_fire_mqtt_message(hass, 'state-topic', '99')
+
+    state = hass.states.get('cover.test')
+    assert STATE_CLOSED == state.state
+
+
+async def test_position_via_template(hass, mqtt_mock):
+    """Test the controlling state via topic."""
+    assert await async_setup_component(hass, cover.DOMAIN, {
+        cover.DOMAIN: {
+            'platform': 'mqtt',
+            'name': 'test',
+            'position_topic': 'get-position-topic',
+            'command_topic': 'command-topic',
+            'qos': 0,
+            'value_template': '{{ (value | multiply(0.01)) | int }}'
+        }
+    })
+
+    state = hass.states.get('cover.test')
+    assert STATE_UNKNOWN == state.state
+
+    async_fire_mqtt_message(hass, 'get-position-topic', '10000')
+
+    state = hass.states.get('cover.test')
+    assert STATE_OPEN == state.state
+
+    async_fire_mqtt_message(hass, 'get-position-topic', '5000')
+
+    state = hass.states.get('cover.test')
+    assert STATE_OPEN == state.state
+
+    async_fire_mqtt_message(hass, 'get-position-topic', '99')
+
+    state = hass.states.get('cover.test')
+    assert STATE_CLOSED == state.state
+
+
+async def test_optimistic_state_change(hass, mqtt_mock):
+    """Test changing state optimistically."""
+    assert await async_setup_component(hass, cover.DOMAIN, {
+        cover.DOMAIN: {
+            'platform': 'mqtt',
+            'name': 'test',
+            'command_topic': 'command-topic',
+            'qos': 0,
+        }
+    })
+
+    state = hass.states.get('cover.test')
+    assert STATE_UNKNOWN == state.state
+    assert state.attributes.get(ATTR_ASSUMED_STATE)
+
+    hass.async_add_job(
```

Comment:
We should never use `hass.async_add_job` in new code, unless we specifically need to handle any type of job.